common/src/main/java/com/google/auto/common/BasicAnnotationProcessor.java

```diff
@@ -505,74 +525,402 @@ private ImmutableSetMultimap<Class<? extends Annotation>, Element> toClassKeyedM
+    }
+
+    @Override
+    protected boolean isAcceptableElementKind(Element element) {
+      return element.getKind().isField();
+    }
+  }
+
+  /**
+   * Represents METHOD and CONSTRUCTOR.
+   *
+   * <p>The {@code equals()} and {@code hashCode()} have been overridden since the {@code toString}
+   * alone is not sufficient to make distinction in all overloaded cases. For example, {@code <C
+   * extends Set<String>> void m(C c) {}}, and {@code <C extends SortedSet<String>> void m(C c) {}}
+   * both have the same toString {@code <C>m(C)} but are valid cases for overloading methods.
+   * Moreover, the needed enclosing type-element information is not included in the toString.
+   *
+   * <p>The executable element is retrieved by saving its enclosing type element, simple name, and
+   * ordinal position in the source code relative to other related overloaded methods, meaning those
+   * with the same simple name. This is possible because according to Java language specification
+   * for {@link TypeElement#getEnclosedElements()}: "As a particular instance of the general
+   * accuracy requirements and the ordering behavior required of this interface, the list of
+   * enclosed elements will be returned to the natural order for the originating source of
+   * information about the type. For example, if the information about the type is originating from
+   * a source file, the elements will be returned in source code order. (However, in that case the
+   * ordering of implicitly declared elements, such as default constructors, is not specified.)"
    *
-   * <p>Packages declared (and annotated) in {@code package-info.java} are tracked as deferred
-   * packages, type elements are tracked directly, and all other elements are tracked via their
-   * nearest enclosing type.
+   * <p>Simple name is saved since comparing the toString as a subtitle is not reliable when at
+   * least one parameter references ERROR, possibly because it is not generated yet. For example,
+   * method {@code void m(SomeGeneratedClass sgc)}, before the generation of {@code
+   * SomeGeneratedClass} has the toString {@code m(SomeGeneratedClass)}; however, after the
+   * generation it will have toString equal to {@code m(test.SomeGeneratedClass)} assuming that the
+   * package name is "test".
    */
-  private static final class ElementName {
-    private enum Kind {
-      PACKAGE_NAME,
-      TYPE_NAME,
+  private static final class ExecutableElementBluePrint extends ElementBluePrint {
+    private final TypeElementBluePrint enclosingTypeElementBluePrint;
+    private final Name simpleName;
+    private final int ordinalOverloadPosition;
+
+    private ExecutableElementBluePrint(Element element) {
+      super(element);
+      TypeElement enclosingTypeElement = getEnclosingType(element);
+      this.enclosingTypeElementBluePrint = new TypeElementBluePrint(enclosingTypeElement);
+      this.simpleName = element.getSimpleName();
+      int ordinalOverloadPosition = 1;
```

Comment:
This might be a bit clearer?

```java
Set<? extends Element> enclosed = enclosingTypeElement.getEnclosedElements();
int ordinalOverloadPosition = 0;
for (int i = 0; i < enclosed.size() && !enclosed.get(i).equals(elt); i++) {
  if (elt.getSimpleName().equals(enclosed.get(i).getSimpleName() && ...) {
    ordinalOverloadPosition = i;
  }
}
```

Then it is 0-origin rather than 1-origin, which I think just means that `<` changes to `<=` in the for-loop below.

Reply:
I've adjusted `ordinalOverloadPosition` to be 0-origin. It's worth noting that, in my implementation, this value represents the ordinal position solely among methods/constructors with the same name, not the order among all enclosed elements, as your code suggests.

While I'm confident there was a specific rationale for this choice, I currently don't recall the exact details. It's plausible that the decision was influenced by the uncertainty in the ordering of implicitly declared elements, such as the default constructor, as outlined in the Java documentation.

Reply:
I see, yes. Although we could imagine that the different processing rounds could return `getEnclosedElements()` in a different order, I think that would be pretty surprising given the specification that you quote. So indeed just storing the position of the element among all enclosed elements might be sufficient? Without needing to have any specific logic for overloading, except perhaps a verification check to ensure that the simple name is still the one we expect.

Alternatively, if you want to be safe and keep the logic as it stands, I think it would be good to get rid of that `Iterator`. I always wince a bit when I see an `Iterator`. 🙂  What about just streaming over the enclosed elements here too, to produce a `List` of elements with the same name, and then using `indexOf` to get `ordinalOverloadPosition`. Perhaps rename that, too.

Reply:
Honest, I feel safer having both name and position within rather than just relying on the ordinal position. This preference stems partly because of the implicit declarations and partly because of practical AST-changing tools like Lombok, as you mentioned before. Since the processing order isn't guaranteed, I feel more secure with my approach. However, if you think sticking to only the ordinal position is the way to go, I'm fine with that.

Regarding the Iterator, I get it. :) How about using a straightforward while loop?

Also, I could use some help with naming the ordinalOverloadPosition. I'm struggling to come up with a self-explanatory name. I changed it to ordinalPositionWithSameName, but I'm not convinced it's the best choice. If you have any suggestions, I'd love to hear them.